eng/pipelines/templates/scripts/validate-changes-on-allowed-path.ps1

```diff
@@ -0,0 +1,64 @@
+<#
+.SYNOPSIS
+
+.DESCRIPTION
+The script is to validate if the diff changes between base and target are within the scope of allowlist.
+
+.PARAMETER BaseBranch
+The base branch where we want to merge changes from. E.g. main branch in docs repo
+
+.PARAMETER TargetBranch
+The target branch where we want to merge changes to. E.g. live branch in docs repo
+
+.PARAMETER AllowlistPath
+The file path of allowlist.
+#>
+param(
+  [Parameter(mandatory=$true)]
+  [string] $BaseBranch,
+  [Parameter(mandatory=$true)]
+  [string] $TargetBranch,
+  [Parameter(mandatory=$true)]
+  [string] $AllowlistPath
+)
+
+# Fetch all branches, e.g. main, live
+Write-Host "git -c user.name=`"azure-sdk`" -c user.email=`"azuresdk@microsoft.com`" fetch origin"
+git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" fetch origin | Write-Host
+
+Write-Host "Check existence on target branch"
+git rev-parse --verify $TargetBranch 2>$null
+if ($LASTEXITCODE -ne 0) {
+  Write-Host "Creating target branch $TargetBranch..."
+  git checkout -b $TargetBranch -t origin/$TargetBranch | Write-Host
+}
+
+# Working branch is base branch, e.g. main
+Write-Host "Check existence on base branch"
+git rev-parse --verify $BaseBranch 2>$null
+if ($LASTEXITCODE -eq 0) {
+  Write-Host "Swtiching to base branch $BaseBranch..."
+  git checkout $BaseBranch | Write-Host
+}
+else {
+  Write-Host "Creating base branch $BaseBranch..."
+  git checkout -b $BaseBranch -t origin/$BaseBranch | Write-Host
+}
+
+# git diff on all files
+$changedFiles = git diff --name-only $TargetBranch
+Write-Host "Here is the total number of the changed files: $changedFiles.Count"
+
+# Get all allowPath, allowPath is the regex of the path
+$allowedPath = Get-Content -path $AllowlistPath
+
+# forloop to see whether there are any changes outside of the allowlist
+Write-Host "Validating the changed files..."
+foreach ($file in $changedFiles) {
+    $matched = $allowedPath | Where-Object {$file -match $_}
```

Comment:
Lets break this into another foreach loop instead of in a piped Where-Object.

Reply:
It is really simple logic. Why do we want complicate the structure?

Reply:
pipeline to a where-object can give you 3 possible outputs which all need to be handled independently. One is null, or a single item, or a list and each are different types and need to be correctly handled. Your if statement after this is not correctly handling those types and as such will fail in some cases. I generally like to avoid using object pipelines in scripts and be very explicit about what I'm doing which is why I like a foreach loop in this case over a Where-Object.

Reply:
I understand it is necessary if we want to handle the errors for the different scenarios. Our cases are super simple like I want to know if the file (not null) matches any allowPath. I can only image the case returns empty or something and these two cases are what we expect. 
1. When does `Where-Object` returns $null to us? Could you share a case?

Right now, we failed at fast. If we change to fail at end, then what we need to do is to add $file into array instead of failing the case. 

Just want to clarify the real concerns here before the change. If it's our team convention, I am definitely willing to make the changes.

Reply:
Tested locally in strict mode. Seems like it indeed returns different things for $null, single item and multiple items. The count function is also different for different cases under strict mode.

Reply:
It is reasonable to make the changes and will switch the script to strict mode at the same time